File: candle/crates/pocket-tts/src/modules/rope.rs
use candle_core::{DType, Result, Tensor};

pub fn apply_rope(
    q: &Tensor,
    k: &Tensor,
    offset: usize,
    max_period: f32,
) -> Result<(Tensor, Tensor)> {
    let (b, t, h, d_full) = q.dims4()?;
    let (_bk, _tk, hk, _dk) = k.dims4()?;

    let d = d_full / 2;
    let dev = q.device();

    // freqs = exp(arange(D/2) * (-log(max_period) * 2 / D))
    let ds = Tensor::arange(0u32, d as u32, dev)?.to_dtype(DType::F32)?;
    let freqs = ds
        .affine((-max_period.ln() * 2.0 / d_full as f32) as f64, 0.0)?
        .exp()?;

    // ts = (arange(T) + offset).view(-1, 1, 1)
    let ts = (Tensor::arange(0u32, t as u32, dev)?
        .to_dtype(DType::F32)?
        .affine(1.0, offset as f64)?)
    .reshape((t, 1, 1))?;

    // freqs * ts -> shape (t, 1, d)
    let freqs_ts = freqs.reshape((1, 1, d))?.broadcast_mul(&ts)?;
    let cos = freqs_ts.cos()?;
    let sin = freqs_ts.sin()?;

    // Reshape q and k to (b, t, h, d, 2)
    let q = q.reshape((b, t, h, d, 2))?;
    let k = k.reshape((b, t, hk, d, 2))?;

    let qr = q.narrow(4, 0, 1)?.squeeze(4)?;
    let qi = q.narrow(4, 1, 1)?.squeeze(4)?;
    let kr = k.narrow(4, 0, 1)?.squeeze(4)?;
    let ki = k.narrow(4, 1, 1)?.squeeze(4)?;

    // qor = qr * cos - qi * sin
    // qoi = qr * sin + qi * cos
    let qor = (qr.broadcast_mul(&cos)? - qi.broadcast_mul(&sin)?)?;
    let qoi = (qr.broadcast_mul(&sin)? + qi.broadcast_mul(&cos)?)?;

    let kor = (kr.broadcast_mul(&cos)? - ki.broadcast_mul(&sin)?)?;
    let koi = (kr.broadcast_mul(&sin)? + ki.broadcast_mul(&cos)?)?;

    let qo = Tensor::stack(&[qor, qoi], 4)?.reshape((b, t, h, d_full))?;
    let ko = Tensor::stack(&[kor, koi], 4)?.reshape((b, t, hk, d_full))?;

    Ok((qo, ko))
}

#[derive(Debug, Clone)]
pub struct RotaryEmbedding {
    max_period: f32,
}

impl RotaryEmbedding {
    pub fn new(max_period: f32) -> Self {
        Self { max_period }
    }

    pub fn forward(&self, q: &Tensor, k: &Tensor, offset: usize) -> Result<(Tensor, Tensor)> {
        apply_rope(q, k, offset, self.max_period)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use candle_core::{Device, Tensor};

    #[test]
    fn test_rope_shape() -> Result<()> {
        let device = Device::Cpu;
        let q = Tensor::zeros((1, 10, 4, 32), DType::F32, &device)?;
        let k = Tensor::zeros((1, 10, 4, 32), DType::F32, &device)?;
        let rope = RotaryEmbedding::new(10000.0);
        let (qo, ko) = rope.forward(&q, &k, 0)?;
        assert_eq!(qo.dims(), &[1, 10, 4, 32]);
        assert_eq!(ko.dims(), &[1, 10, 4, 32]);
        Ok(())
    }
}

File: pocket_tts/modules/rope.py
import math

import torch
from torch import nn


def apply_rope(q, k, offset, max_period):
    """
    Args:
        q (torch.Tensor): Queries, shape `[B, T, H, D]`.
        k (torch.Tensor): Keys, shape `[B, T, H, D]`.
        offset (int): Current offset, e.g. when streaming.
        max_period (float): Maximum period for the cos and sin.
    """

    B, T, H, D = q.shape
    Bk, Tk, Hk, Dk = k.shape
    if (Bk, Tk, Dk) != (B, T, D):
        raise ValueError(
            f"Query and Key shapes must match except for head dimension: q={(B, T, D)}, k={(Bk, Tk, Dk)}"
        )
    if D <= 0:
        raise ValueError(f"Dimension D must be positive, got {D}")
    if D % 2 != 0:
        raise ValueError(f"Dimension D must be even, got {D}")
    if max_period <= 0:
        raise ValueError(f"max_period must be positive, got {max_period}")

    ds = torch.arange(D // 2, device=q.device, dtype=torch.float32)
    freqs = torch.exp(ds * (-math.log(max_period) * 2 / D))

    # could be optimized in one call
    ts = torch.arange(T, device=q.device, dtype=torch.float32)
    ts += offset
    ts = ts.view(-1, 1, 1)

    q = q.view(B, T, H, D // 2, 2)
    k = k.view(B, T, Hk, D // 2, 2)

    # convention is `r` suffix is real part, `i` is imaginary.
    qr = q[..., 0].float()
    qi = q[..., 1].float()

    kr = k[..., 0].float()
    ki = k[..., 1].float()

    rotr = torch.cos(freqs * ts)
    roti = torch.sin(freqs * ts)
    qor = qr * rotr - qi * roti
    qoi = qr * roti + qi * rotr

    kor = kr * rotr - ki * roti
    koi = kr * roti + ki * rotr

    dtype = q.dtype
    qo = torch.stack([qor.to(dtype), qoi.to(dtype)], dim=-1)
    ko = torch.stack([kor.to(dtype), koi.to(dtype)], dim=-1)

    return qo.view(B, T, H, D), ko.view(B, T, Hk, D)


class RotaryEmbedding(nn.Module):
    """Rotary positional embedding (RoPE) from [Su et al 2022](https://arxiv.org/abs/2104.09864).

    Args:
        max_period (float): Maximum period of the rotation frequencies.
    """

    def __init__(self, max_period=10000.0):
        super().__init__()
        self.max_period = max_period

    def forward(self, q, k, offset):
        """Apply rope rotation to query or key tensor."""
        mp = q.new_full((), self.max_period)
        return apply_rope(q, k, offset, mp)
